### ISSUE TITLE
Separate Optimizers for each Head in the MultiTaskModel

### ISSUE THREAD

**Issue by jbrry** [allenai/allennlp#4911]
**Is your feature request related to a problem? Please describe.**
I'm using the `multitask` model which consists of multiple `heads`. At the moment, the loss from each head is summed together and is then jointly optimized using a single optimizer. I currently have a head where the input is only a word-level feature and another head where the input is only characters. The heads have comparable performance when I use a pre-trained word embedding or a randomly initialized embedding, e.g. the word and character heads achieve `0.837` and `0.819` accuracy respectively. When I changed the word-level feature to using the output of a pre-trained transformer, the scores of the heads change dramatically to `0.902` and `0.210` for word and character models respectively. In this case, the head which uses an embedding from the pretrained transformer starts to dominate.

**Describe the solution you'd like**
There currently is an option `loss_weights` which controls the weight to apply to the loss from each head. However, the [work I am replicating](https://arxiv.org/pdf/1805.08237.pdf) uses separate optimizers for each head and they found this to be most effective: 

> Separate optimization outperformed joint optimization by up to 2.1 percent absolute [...] We hypothesize that separately training the models forces each sub model (word and character) to be strong enough to make high accuracy predictions and in some sense serves as a regularizer in the same way that dropout does for individual neurons.

This approach might be better than manipulating the `loss_weights` values, which might differ between tasks and would thus require some level of intuition to properly control. Additionally, separate optimizers would enable each head to minimize its own loss independently which could result in individually stronger heads.

Admittedly, I do not have that much experience with the `Trainer` code but I imagine a crude solution would be to create my own `Trainer`, e.g. `gradient_descent_separate_optimizers` and change the type of the optimizer from `optimizer : torch.nn.Optimizer` to `optimizers : Dict[str, torch.nn.Optimizer]`, where the optimizer is keyed on the head name. Then I could mostly duplicate some of the logic in the `gradient_descent` `Trainer` but instead of calling the individual optimizer, e.g. `self.optimizer.step()`, the head-specific optimizer would be called, optimizing the loss for each head.

I'm just wondering would this be the best way of going about this? I'd be happy to hear if you have any suggestions or advice for implementing this or if you would think it would be a good idea to have as part of the multitask code. Thanks!

**Comment by dirkgr** [comment]
The concerns
--

It seems theoretically unsound to do this unless the backbone is fixed. This is the problem I see: Most optimizers will keep track of statistics like momentum for each parameter. If you then run another optimizer over the same weights, it will move the weights and thus invalidate the momentum, but the first optimizer is none the wiser and just keeps on trucking with the momentum values it has.

From an engineering perspective, there is a risk that in PyTorch the optimizers will try to write their momentum values into the same state variables and overwrite each other's internal state, though I'd want to confirm that this is really a problem before getting too concerned about it.

You can already configure hyper parameters like the learning rate per task, or per weight (using parameter groups).

The solution
--

All of this said, maybe your backbone is fixed, or you just want to replicate someone else's work so you don't care whether it's sound or not, so you need to do it anyways, how to do it? I think there'd be a number of things to do to make this a real feature:
 1. Change `MultitaskModel` so that it outputs per-task losses in addition to the sum of the loss.
 2. Write a variant of `GradientDescentTrainer` that, like you said, keeps a dictionary of optimizers, and uses the separate losses from step 1 instead of the sum of the losses.
 2. Give `MultitaskModel` a parameter `freeze_backbone: bool = False`, to make it easy to do the right thing. Ideally we print a warning if we detect that multiple optimizers are operating on the same weights. If we find that the optimizers do actually clobber each other's internal state, we should throw an exception instead of just a warning.

Would you be willing to make a PR for this? I'd be happy to review it.

Please note that I have changed the multitask code a bit in https://github.com/allenai/allennlp/pull/4898. I'm hoping to merge that today or tomorrow.

**Comment by jbrry** [comment]
Thanks for your insights Dirk. I agree that one needs to be cautious about how the backbone or any other shared components are optimized, e.g. you do not want multiple optimizers operating on the same set of parameters as you have mentioned.

I think in theory you could have head-specific optimizers which optimize the parameters related to that head. You could then also include a shared optimizer which combines the loss of the heads and optimizes "shared" parameters, e.g. a shared encoder in the backbone.

In my particular case, my backbone embeds and encodes words and characters separately which are then passed to their respective heads as input features for a parsing/classification task. So I think what I would need to do is create parameter groups which are relevant to each head, so that the word parameters in the backbone are optimized using the word head loss and the character parameters in the backbone are optimized using the loss from the character head. To create these parameter groups, I may need to change some of the code in [make_parameter_groups (training/optimizers.py)](https://github.com/allenai/allennlp/blob/f11a0f469ceed12f4822cb76cfffddb7c7304ba1/allennlp/training/optimizers.py#L33) so that groupings can be made which match the right parameters from the head and backbone, keyed on a common name which will also be used for the optimizer mapping. These names can already be inferred from the named parameters of the model. For instance, I am using `word` and `token_character` as names for heads and components in my setup. Addmittedly, these names are a bit ambiguous but in the example in `MultiTaskExample`, they would be the head names like: `"vqa", "gqa", "ve"`.

```
# backbone parameters
_backbone._word_embedder.token_embedder_tokens.weight
_backbone._token_character_embedder.token_embedder_token_characters.weight

# head parameters
_heads.word.head_tag_feedforward._linear_layers.0.weight
_heads.token_character.head_tag_feedforward._linear_layers.0.weight
```

Thank you for the suggestions, I will try to implement these. Yes I'd be happy to submit a PR, thanks for volunteering to review it! Should I work offline for a bit and submit a PR after I have made some initial progress or would you prefer I make a PR soon so you can track the changes as they come in? Also, no stress with this if you find it over-complicates things but I'd be happy to work on it anyway and then a final decision could be made if it is beneficial to have in the library.

**Comment by dirkgr** [comment]
I have looked at this again, and I think there is a better solution: Write a custom optimizer that defers to other optimizers.

When optimizers are initialized, they already get a set of parameters that they work on. So the big optimizer could take an argument much like the existing optimizer takes `"parameter_groups"`. For each regex you specify, you can set a different optimizer. `optimizer.step()` then performs the step on all the sub-optimizers.

In the config, it would look something like this:
```
    "optimizer": {
      "type": "regex",
      "optimizers": [
        ["^backbone\\.", {
          "type": "huggingface_adamw",
          "lr": 4e-5,
          "correct_bias": true,
          "weight_decay": 0.01,
        }],
        ["^head1\\.", {
          "type": "adagrad",
          ...
        }],
        ["^head2\\.", {
          "type": "sgd",
          ...
        }]
      ]
    }
```

This might also need a default optimizer, for when none of the regexes match, or maybe it should just throw an exception if that happens. There should be a check against parameters that match multiple optimizers.

In this way, the functionality is completely separate from the multitask thing and could be used in other contexts as well. Also, you don't have to mess with the trainer at all, which is always a big plus. Maybe for convenience we could also add a `MultitaskOptimizer` that just defers to the `RegexOptimizer`, so you don't have to write regexes in your case.

**Comment by jbrry** [comment]
Hi Dirk, just to give you an update on this:

I created a RegexOptimizer [here](https://github.com/jbrry/allennlp/blob/98ed2a4a0b2e889741329ad7b7a060f1c33d442e/allennlp/training/optimizers.py#L200). Like in your example above, it takes in an argument `optimizers` which specifies the regexes to use and the corresponding optimizer. It creates an object `self._grouped_optimizers` which is a dictionary mapping some metadata `name` to the optimizer.

This object can be accessed in [`trainer.py`](https://github.com/jbrry/allennlp/blob/separate_optimizers/allennlp/training/trainer.py). I don't think I was able to leave the code of `gradient_descent` trainer unchanged. What I have done so far is that any time the optimizer is used, there is first a check to see whether the optimizer is a dictionary. If so, there is just some simple recursion to access the optimizers and if not, the code remains the same. I have confirmed that I am able to train with multiple optimizers, e.g. by adding the below to my config file (but I still need to do more checks about restoring `state_dicts` etc.):

```
    "optimizer": {
      "type": "regex",
      "optimizer_ignore_keys": ["params", "name"],
      "default_optimizer_kwargs": {
        "type": "adam",
        "lr": 0.001,
        "betas": [0.9, 0.9],
      },
      "optimizers": [
        [["text_field_embedder"], {
          "type": "adam",
          "name": "embedder",
          "lr": 0.001,
          "betas" : [0.9, 0.9],
        }],
      ]
    }    
```

One or two things popped up which I still need to change:
1. Normally, when optimizers are initialized, the parameters for the optimizer are created with the `__init__` method for each optimizer, e.g.: `params=make_parameter_groups(model_parameters, parameter_groups)`. In my implementation I am calling the `make_parameter_groups` function once and then deferring to the optimizers. I found this was easier than doing it separately each time as we would need to keep track of changes between parameter and default groups each time.
2. When initializing the `Optimizer` object in the `Trainer`, if the optimizer is not a typical `torch.optim.Optimizer`, there is an `AttributeError` that it doesn't have `param_groups` so I added a simple try, except block to check if we have a `_grouped_optimizers` object first and if not, we fall back to using a single `Optimizer`.

**Comment by dirkgr** [comment]
I thought the `RegexOptimizer` would override `step()` (and `zero_grad()`, `state_dict()`, and `load_state_dict()`) to call its inner optimizers, but behave like a normal optimizer on the outside. Did you try this and it didn't work?

**Comment by jbrry** [comment]
> I thought the RegexOptimizer would override step() (and zero_grad(), state_dict(), and load_state_dict()) to call its inner optimizers, but behave like a normal optimizer on the outside.

Sorry, I kind of overlooked how the `RegexOptimizer` could implement these methods so I didn't try it. This would be a better solution so I will try to implement these inside of the `RegexOptimizer` class. I can then update you with how that goes!

**Comment by jbrry** [comment]
I was able to implement the methods, `step()`, `state_dict()` and `load_state_dict()` in the [`RegexOptimizer`](https://github.com/jbrry/allennlp/blob/40aef6017f8c83680efd38952fe0765481970ad7/allennlp/training/optimizers.py#L258). I also tried training with the `--recover` flag to make sure `load_state_dict()` works ok, which it does. This means there are now no changes to the code in `trainer.py`.

I was not able to implement `zero_grad()` because in the [`Trainer`](https://github.com/jbrry/allennlp/blob/40aef6017f8c83680efd38952fe0765481970ad7/allennlp/training/trainer.py#L606), parameter gradients are manually set to `None` (this is meant to be a bit quicker than calling `zero_grad()`:
```
for param_group in self.optimizer.param_groups:
    for p in param_group["params"]:
        p.grad = None
```

I was worried that operating on the parameters of `param_group` in `self.optimizer.param_groups` would not change the parameters which are part of a `param_group` of the individual sub-optimizers, (i.e. in `self.optimizer._grouped_optimizers`) but I was able to confirm that the gradients of all sub-optimizer parameters are altered by the above loop with the check below. I can't confirm exactly why this occurs yet as I thought they may be separate instantiations of the same parameters but it should mean that we don't need to recurse over each optimizer in `self.optimizer._grouped_optimizers` in the `Trainer` like I had been doing before.

```
# Check that the parameters of the sub-optimizers were also changed:
# This is purely for debugging purposes.
for name, optimizer in self.optimizer._grouped_optimizers.items():
    for param_group in optimizer.param_groups:
        for p in param_group["params"]:
            assert p.grad == None
```

Similarly, the `GradientScaler` operates on a single optimizer, so I believe any changes the `scaler` makes to `self.optimizer.param_groups` (i.e. scaling/unscaling) will be applied to the parameters of the sub-optimizers as well but I would perhaps wait for individual unit tests to confirm this.

Apologies if the code is a little immature to be discussing right now but I wanted to let give you an update. I can try to use it for my purposes described above to see how that goes and implement the separate losses / `Backbone` changes too and can update you when I've made some progress.

**Comment by dirkgr** [comment]
I think what's going on is that you have `self.param_groups` in the parent optimizer that points to all the parameters, and `self.param_groups` in the child optimizers pointing only to their own parameters. Every parameter is pointed to twice. And from the looks of it, that is fine. I wouldn't mind getting rid of the `p.grad = None` optimization, but doing that with the scaler would be harder.

Let me know when the code is ready for a review! I'd be happy to give it a look.

**Comment by jbrry** [comment]
Thanks, I'll look into ways I can avoid any duplication where possible.

Regarding `p.grad = None`, it seems that the [`zero_grad()`](https://github.com/pytorch/pytorch/blob/9f6e0de5485900eed3d5e2438b4797e8744ce088/torch/optim/optimizer.py#L193) method in the base Optimizer class in PyTorch has a `set_to_none` argument which enables either setting the gradient to zero or `None`, so this could actually provide the best of both worlds.

Thanks a lot Dirk, will do!

**Comment by dirkgr** [comment]
I don't think there is a problem with pointing to the parameter tensors twice. That sounds like a really elegant solution actually. If that's really the situation, I don't think it needs fixing.

**Comment by jbrry** [comment]
I have updated some of the code which now allows for multiple optimizers and losses in this commit: https://github.com/jbrry/allennlp/commit/de9c6ec1d82c302ea69abf3d0c7c49123af2ae2d. I am working on the branch [`separate_optimizers_and_losses`](https://github.com/jbrry/allennlp/tree/separate_optimizers_and_losses) in my AllenNLP fork.

Regarding the `RegexOptimizer`, there are some changes: The `optimizers` argument is now a list of optimizers, with optimizer keyword arguments as well as an additional `name` parameter. Then `parameter_groups` is used as normal (i.e. for the regexes and the optimizer group options) but also a `name` parameter is smuggled in which matches this parameter group to the appropriate optimizer and can be used to override optimizer options as well. If no `name` is given, it gets assigned to the `default` group. Having both `optimizers` and `parameter_groups` as parameters made it a bit easier for the case where you would want to set non-default optimizer options for certain groups.

When the [sub-optimizers are initialized](https://github.com/jbrry/allennlp/blob/separate_optimizers_and_losses/allennlp/training/optimizers.py#L244), I am passing `parameter_groups` as a boolean `False` which makes `make_parameter_groups` a no-op. This is because we have already called this function once inside `RegexOptimizer` to make our parameter groups but the `make_parameter_groups` function is called by each `Optimizer` class when being initialized. This seems a little bit hacky but it was one of the more straightforward ways of not having to change any of the code for the individual optimizers. Here is how it looks in a config:

```
      "optimizer": {
        "type": "regex",
        "optimizers": [
        {
          "name": "default",
          "type": "adam",
          "lr": 0.001,
          "betas" : [0.9, 0.9],
        },
        {
          "name": "character",
          "type": "adam",
          "lr": 0.001,
          "betas" : [0.9, 0.9],
        },
        {
          "name": "meta",
          "type": "adam",
          "lr": 0.001,
          "betas" : [0.9, 0.9],
        },
        {
          "name": "word",
          "type": "adam",
          "lr": 0.001,
          "betas" : [0.9, 0.9],
        },
      ],
      "parameter_groups": [
        [["meta"], {
          "name": "meta",
          "lr": 0.001,
          "betas" : [0.9, 0.99], // you can override the main optimizer options here as normal
        }],
        [["character"], {
          "name": "character",
        }],
        [["word"], {
          "name": "word",
        }],
      ],
    }
```

Regarding the optimizer's parameters, I believe they are pointed to when the `RegexOptimizer` is initialized, e.g. can be accessed in `self.optimizer.param_groups`. They are also pointed to in `self.optimizer._grouped_optimizers` so I believe what's going on is that they are both different objects pointing to the same parameters, so when the parameters are modified in one they will also be modified in the other. I have confirmed that `id()` on all of the parameters within the two objects are the same.

I changed some of the code in the `Trainer` to [operate on multiple losses](https://github.com/jbrry/allennlp/blob/separate_optimizers_and_losses/allennlp/training/trainer.py#L709). To facilitate this, the `MultiTaskModel` has an option to [create a dictionary entry](https://github.com/jbrry/allennlp/blob/de9c6ec1d82c302ea69abf3d0c7c49123af2ae2d/allennlp/models/multitask.py#L153): `outputs["task_losses"]` which matches a `f"{head_name}_loss"` to its `loss` and these loss values are accessed by the `Trainer` if the `multiple_losses` argument is passed. If it is, we recursively [call `backward()`](https://github.com/jbrry/allennlp/blob/de9c6ec1d82c302ea69abf3d0c7c49123af2ae2d/allennlp/training/trainer.py#L744) on each task's loss.

I was able to train my multitask model with separate losses and optimizers for the purposes of my original post. If you still think this is a useful feature, I'd be happy to polish it up a bit so that it's suitable for a PR and could implement some tests but there is no pressure if you find it overcomplicates things, especially as it requires some changes in the `gradient_descent` trainer to operate on multiple losses.

**Comment by dirkgr** [comment]
Thank you for doing this! I'd love to have the `RegexOptimizer` in a separate PR. I think that would be a cool addition to the optimizers.

I am not yet sold on multiple losses. What's the problem you're solving that way?

Either way, those should be separate PRs, since they are separate features, and at least `RegexOptimizer` is quite large as it is.

**Comment by jbrry** [comment]
Ok great! I will submit a PR soon. Should I add a few tests to `https://github.com/allenai/allennlp/blob/main/tests/training/optimizer_test.py` to make sure everything's working as expected as part of the PR?

> I am not yet sold on multiple losses. What's the problem you're solving that way?

My thinking was that having separate losses would mean that each optimizer would only know about the gradients of a particular loss so it wouldn't be influenced by the gradients in another loss. Looking into it more, I don't think this matters anyway because if a parameter is not involved in the computation of one particular loss, then its gradients won't be affected by that loss, even if the losses are summed together at a later point. I just wrote a small example in the below file to check if that was the case and it seems to be, e.g. `x2`'s gradients stay the same if `backward()` is called separately on each loss or if the losses are summed together in the second example, so I don't think there is any reason to have the separate losses now (which is a good thing!).

[compare_separate_and_summed_losses.txt](https://github.com/allenai/allennlp/files/5947719/compare_separate_and_summed_losses.txt)

**Comment by dirkgr** [comment]
> Should I add a few tests

Yes please :-)

**Comment by jbrry** [comment]
Thanks for all your help with this @dirkgr and for spending the extra time improving it!
